### PR TITLE
Edit user profile

### LINKED ISSUE #37: View User Profile

### LINKED ISSUE #61: Update User Profile

### PR DESCRIPTION
**Pull request by carriezhengjr**
Fixes #37
Fixes #61 

Frontend:
- Username and email will be pre-filled with initial user data.
- If modal is closed without saving the edited data, the initial user data will be shown again when user reopens modal (instead of showing edited data).
- For Languages:
  - Dropdown appears when user clicks on the input field.
  - Autocompletes when user types.

Backend:
- Able to update username, email, languages.
  - Successful update: page will automatically show updated details.
  - Messages to prompt user when username or email is empty, or when it already exists.

Screenshot 2023-10-06 at 22 14 28

Screenshot 2023-10-06 at 22 15 14

### CODE REVIEW

File: frontend/src/pages/users/EditProfile.tsx [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+                      value={language}
+                      style={{ marginTop: 4, marginBottom: 4 }}
+                      _selected={{ bg: useColorModeValue('blackAlpha.50', 'whiteAlpha.50'), color: 'gray.500' }}
+                      _focus={{ bg: useColorModeValue('blackAlpha.100', 'whiteAlpha.100') }}
+                    >
+                      {language}
+                    </AutoCompleteItem>
+                  ))}
+                </AutoCompleteList>
+              </AutoComplete>
```
Comment (seelengxd) [inline frontend/src/pages/users/EditProfile.tsx]: autocomplete has been just refactored to a separate component in `frontend/src/components/questions/QuestionCategoryAutocomplete.tsx`

oh wait my bad i just realised this was for languages, ignore this oops

but there is a new <MultiSelect /> component the rest use

File: frontend/src/pages/users/ViewProfile.tsx [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+              {user?.role}
+            </Text>
+          </HStack>
+          <HStack spacing={5} align="center" mt={4}>
+            <FaCode fontSize="24px" />
+            <Text>
+              <span style={{ fontWeight: 'bold' }}>Languages: </span>
+              {(user?.languages?.length ?? 0) > 0 ? user?.languages?.map((lang) => lang.language).join(', ') : 'None'}
+            </Text>
+          </HStack>
```
Comment (seelengxd) [inline frontend/src/pages/users/ViewProfile.tsx]: nit: icon seems to shrink if there are many languages
image

File: backend/user-service/src/controllers/authController.ts [hunk 3]
```diff
@@ -375,3 +400,83 @@ export async function updateAccessToken(req: Request, res: Response) {
...
+        },
+        data: {
+          username: username,
+          email: email,
+          languages: {
+            set: [],
+            connect: languageIds.map((id) => ({ id: id })),
+          },
+        },
+      });
```
Comment (seelengxd) [inline backend/user-service/src/controllers/authController.ts:460]: nit: hashed password getting returned to user
image

**Review by seelengxd — COMMENTED** [review]
note that if this pr is merged as is, future microservices cannot rely on any of the user data in the jwt except user id due to not updating jwt 👍 

but currently none of the other microservices use anything other than id (?)

**Review by seelengxd — APPROVED** [review]

File: backend/user-service/src/controllers/authController.ts (hunk 3 above)
Comment (seelengxd) [reply backend/user-service/src/controllers/authController.ts:460]: ~wait my bad sorry i think i was running something outdated~ wait no its still there in the response but hmm 🤷‍♂️

**Comment by juslam19** [comment]
If I am not wrong, I mentioned that there is a backend API for JWT to update both tokens for access and refresh. @seelengxd

**Comment by juslam19** [comment]
This edit user profile could use that API in user-service

**Pull request merged**